clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -776,32 +778,23 @@ public PartitionRecords(long fetchOffset, TopicPartition partition, List<Consume
         }
 
         private boolean isEmpty() {
-            return records == null || records.isEmpty();
+            return records == null || position >= records.size();
         }
 
         private void discard() {
             this.records = null;
         }
 
         private List<ConsumerRecord<K, V>> take(int n) {
-            if (records == null)
-                return new ArrayList<>();
+            if (isEmpty())
+                return Collections.emptyList();
 
-            if (n >= records.size()) {
-                List<ConsumerRecord<K, V>> res = this.records;
-                this.records = null;
-                return res;
-            }
-
-            List<ConsumerRecord<K, V>> res = new ArrayList<>(n);
-            Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
-            for (int i = 0; i < n; i++) {
-                res.add(iterator.next());
-                iterator.remove();
-            }
+            int limit = Math.min(records.size(), position + n);
+            List<ConsumerRecord<K, V>> res = Collections.unmodifiableList(records.subList(position, limit));
 
-            if (iterator.hasNext())
-                this.fetchOffset = iterator.next().offset();
+            position = limit;
```

Comment:
Nevermind, I was thinking of `position += n`, but `limit` is actually the result of that (bounded by records.size).